flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierProducerAgent.java

```diff
@@ -0,0 +1,175 @@
+
+    /**
+     * Whether a subpartition has been registered to the netty service. The array index is
+     * corresponding to the subpartition id.
+     */
+    private final boolean[] nettyServiceRegistered;
+
+    private final MemoryTierSubpartitionProducerAgent[] subpartitionProducerAgents;
+
+    public MemoryTierProducerAgent(
+            TieredStoragePartitionId partitionId,
+            int numSubpartitions,
+            int bufferSize,
+            int segmentSizeBytes,
+            boolean isBroadcastOnly,
+            TieredStorageMemoryManager memoryManager,
+            TieredStorageNettyService nettyService,
+            TieredStorageResourceRegistry resourceRegistry) {
+        checkArgument(segmentSizeBytes >= bufferSize, "One segment contains at least one buffer.");
+
+        this.segmentSizeBytes = segmentSizeBytes;
+        this.segmentSizeBuffers = segmentSizeBytes / bufferSize;
+        this.isBroadcastOnly = isBroadcastOnly;
+        this.memoryManager = memoryManager;
+        this.numSubpartitionEmitBytes = new int[numSubpartitions];
+        this.nettyServiceRegistered = new boolean[numSubpartitions];
+        this.subpartitionProducerAgents = new MemoryTierSubpartitionProducerAgent[numSubpartitions];
+
+        Arrays.fill(numSubpartitionEmitBytes, 0);
+        nettyService.registerProducer(partitionId, this);
+        for (int subpartitionId = 0; subpartitionId < numSubpartitions; ++subpartitionId) {
+            subpartitionProducerAgents[subpartitionId] =
+                    new MemoryTierSubpartitionProducerAgent(subpartitionId, nettyService);
+        }
+        resourceRegistry.registerResource(partitionId, this::releaseResources);
+    }
+
+    @Override
+    public boolean tryStartNewSegment(TieredStorageSubpartitionId subpartitionId, int segmentId) {
+        if (isBroadcastOnly) {
+            return false;
+        }
+        boolean canStartNewSegment =
+                isSubpartitionRegistered(subpartitionId)
+                        && (memoryManager.getMaxNonReclaimableBuffers(this)
+                                        - memoryManager.numOwnerRequestedBuffer(this))
+                                > segmentSizeBuffers;
+        if (canStartNewSegment) {
+            getSubpartitionProducerAgent(subpartitionId.getSubpartitionId())
+                    .addSegmentId(segmentId);
+        }
+        return canStartNewSegment;
+    }
+
+    @Override
+    public boolean tryWrite(TieredStorageSubpartitionId subpartitionId, Buffer finishedBuffer) {
+        int subpartitionIndex = subpartitionId.getSubpartitionId();
+        if (numSubpartitionEmitBytes[subpartitionIndex] != 0
+                && numSubpartitionEmitBytes[subpartitionIndex] + finishedBuffer.readableBytes()
+                        > segmentSizeBytes) {
+            appendEndOfSegmentEvent(subpartitionIndex);
+            numSubpartitionEmitBytes[subpartitionIndex] = 0;
+            return false;
+        }
```

Comment:
The name `numSubpartitionEmitBytes` is misleading. This is the emitted bytes only for the current segment of the subpartition.

Reply:
Rename this to `currentSubpartitionWriteBuffers`.